File: src/main/java/ch/obermuhlner/astro/image/ImageWriter.java
package ch.obermuhlner.astro.image;

import mil.nga.tiff.TiffWriter;

import javax.imageio.ImageIO;
import java.awt.image.*;
import java.io.File;
import java.io.IOException;

public class ImageWriter {

  public static void write(DoubleImage image, File output) throws IOException {
    String name = output.getName();

    for (ImageFormat format : ImageFormat.values()) {
      for (String extension : format.getExtensions()) {
        if (name.length() > extension.length() && name.substring(name.length() - extension.length()).equalsIgnoreCase(extension)) {
          write(image, output, format);
          return;
        }
      }
    }

    write(image, output, ImageFormat.TIF);
  }

  public static void write(DoubleImage image, File output, ImageFormat format) throws IOException {
    if (image instanceof TiffDoubleImage) {
      if (format == ImageFormat.TIF) {
        TiffDoubleImage tiffDoubleImage = (TiffDoubleImage) image;
        TiffWriter.writeTiff(output, tiffDoubleImage.tiffImage);
        return;
      }

      int width = image.getWidth();
      int height = image.getHeight();
      double[] rgb = new double[3];
      DoubleImage imageCopy = ImageCreator.create(width, height, ImageQuality.Standard);
      for (int y = 0; y < height; y++) {
        for (int x = 0; x < width; x++) {
          image.getPixel(x, y, ColorModel.RGB, rgb);
          imageCopy.setPixel(x, y, ColorModel.RGB, rgb);
        }
      }
      image = imageCopy;
    }

    if (!(image instanceof AwtBufferedDoubleImage)) {
      DoubleImage temp = new AwtBufferedDoubleImage(new BufferedImage(
          image.getWidth(),
          image.getHeight(),
          BufferedImage.TYPE_INT_RGB
      ));
      ImageUtil.copyPixels(image, 0, 0, temp, 0, 0, image.getWidth(), image.getHeight(), ColorModel.RGB);
      image = temp;
    }

    AwtBufferedDoubleImage bufferedDoubleImage = ((AwtBufferedDoubleImage) image);
    ImageIO.write(bufferedDoubleImage.image, format.name(), output);
    return;
  }
}

File: src/main/java/ch/obermuhlner/astro/image/ImageWriter.kt
package ch.obermuhlner.astro.image

import ch.obermuhlner.astro.image.color.ColorModel
import mil.nga.tiff.TiffWriter
import java.awt.image.BufferedImage
import java.io.*
import javax.imageio.ImageIO

object ImageWriter {
    @Throws(IOException::class)
    fun write(image: DoubleImage, output: File) {
        val name = output.name
        for (format in ImageFormat.values()) {
            for (extension in format.extensions) {
                if (name.length > extension.length && name.substring(name.length - extension.length).equals(extension, ignoreCase = true)) {
                    write(image, output, format)
                    return
                }
            }
        }
        write(image, output, ImageFormat.TIF)
    }

    @Throws(IOException::class)
    fun write(image: DoubleImage, output: File, format: ImageFormat) {
        var image = image
        if (image is TiffDoubleImage) {
            if (format == ImageFormat.TIF) {
                TiffWriter.writeTiff(output, image.tiffImage)
                return
            }
            val width = image.width
            val height = image.height
            val rgb = DoubleArray(3)
            val imageCopy = ImageCreator.create(width, height, ImageQuality.Standard)
            for (y in 0 until height) {
                for (x in 0 until width) {
                    image.getPixel(x, y, ColorModel.RGB, rgb)
                    imageCopy.setPixel(x, y, ColorModel.RGB, rgb)
                }
            }
            image = imageCopy
        }
        if (image !is AwtBufferedDoubleImage) {
            val temp: DoubleImage = AwtBufferedDoubleImage(BufferedImage(
                    image.width,
                    image.height,
                    BufferedImage.TYPE_INT_RGB
            ))
            temp.setPixels(0, 0, image, 0, 0, image.width, image.height, ColorModel.RGB, null)
            image = temp
        }
        val bufferedDoubleImage = image as AwtBufferedDoubleImage
        ImageIO.write(bufferedDoubleImage.image, format.name, output)
    }
}